server/src/main/java/io/spine/server/entity/storage/EntityRecordStorage.java

```diff
@@ -0,0 +1,339 @@
+    public boolean delete(I id) {
+        return super.delete(id);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Overrides the parent method in sake of the covariance of the returned value.
+     */
+    @Internal
+    @Override
+    @SuppressWarnings("unchecked")  // Guaranteed by the generic declaration of `EntityRecordSpec`.
+    public final EntityRecordSpec<I, S, ?> recordSpec() {
+        return (EntityRecordSpec<I, S, ?>) super.recordSpec();
+    }
+
+    /**
+     * Transforms the iterator over the entity records into the iterator over their identifiers.
+     *
+     * @param iterator
+     *         source iterator
+     * @return iterator over the identifiers of the passed {@code EntityRecord}s
+     */
+    @SuppressWarnings("unchecked")
+    private Iterator<I> asIdStream(Iterator<EntityRecord> iterator) {
+        ImmutableList<I> ids = stream(iterator)
+                .map(record -> Identifier.unpack(record.getEntityId()))
+                .map(id -> (I) id)
+                .collect(toImmutableList());
+        return ids.iterator();
+    }
+
+    @SuppressWarnings("unchecked")  // ensured by the `EntityClass` definition.
+    private Class<S> stateType() {
+        return (Class<S>) recordSpec().entityClass()
+                                      .stateClass();
+    }
+
+
+    @SuppressWarnings("unchecked")  // ensured by the `Entity` declaration.
+    private Class<I> idType() {
+        return (Class<I>) recordSpec().entityClass().idClass();
+    }
+
+    private static boolean hasNoIds(Query<?, ?> query) {
+        return query.subject()
+                    .id()
+                    .values()
+                    .isEmpty();
+    }
+
+    private static <I> boolean hasNoLifecycleCols(Query<I, ?> query) {
+        Subject<I, ?> subject = query.subject();
+        ImmutableList<? extends QueryPredicate<?>> predicates = subject.predicates();
+        for (QueryPredicate<?> predicate : predicates) {
+            boolean hasColumn = predicate.parameters()
+                                         .stream()
+                                         .anyMatch((param) -> isLifecycleColumn(param.column()));
+            if (hasColumn) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private RecordQuery<I, EntityRecord> onlyActive(RecordQuery<I, EntityRecord> query) {
+        RecordQuery<I, EntityRecord> result = query;
+        if (hasNoIds(query) && hasNoLifecycleCols(query)) {
+            RecordQueryBuilder<I, EntityRecord> builder = query.toBuilder();
+            if(hasArchivedColumn) {
+                builder.where(archived.asRecordColumn(Boolean.class)).is(false);
+            }
+            if(hasDeletedColumn) {
+                builder.where(deleted.asRecordColumn(Boolean.class)).is(false);
+            }
+            result = builder.build();
+        }
+        return result;
+    }
+
+    private static boolean isLifecycleColumn(Column<?, ?> column) {
```

Comment:
I think we should consider having `LifecycleColumns` or something like that. Yes, we have only two of them, but just in case we need yet another one in the future, we'd need to gather the knowledge of such columns in may places in the code. It would be good to consolidate it under a class.

For example, `EntityRecordWithColumns.create(I id, EntityRecord record)` deals with two columns placing them into a `Map`. This class has at least two places where it checks for `archived` and then `deleted` columns, see constructor and `onlyActive()`. Can we consolidate this knowledge under one class?